<!--Sample Summary Page - displays variants, WC, history , reports etc   -->

{% extends 'VariantDatabase/base.html' %}



  {% block content %}

 <style>
                .thumbnail
                {
                    text-decoration:none;
                    color:black;
                    font-weight:bold;
                }
                .thumbnail span  /*CSS for enlarged image*/
                {
                    visibility: hidden;
                    position: absolute;
                    padding: 5px;
                    border: 1px solid #000;
                    background-color: #e5e5e5;
                }
                .thumbnail:hover span  /*CSS for enlarged image on hover*/
                {
                    visibility: visible;
                    left: 600px;
                    top: 750px;
                }
                .imgs td
                {
                    vertical-align:middle;
                    padding: 0.5em;
                    border: 1px solid black;
                }
                table.imgs
                {
                    border-collapse:collapse;
                }
                .nums th
                {
                    text-align: left;
                }
                table.nums
                {
                    margin-top: 1em;
                    margin-left:20px;
                    border: 1px dotted #83A4C3;
                    background-color: #F5F5F5;
                    padding: 0.5em;
                }
                .pad { padding-left:1em; vertical-align:top; }
                .right { text-align:right; padding-left:1em; }


				.header-fixed {
				    width: 100% 
				}

				.header-fixed > thead,
				.header-fixed > tbody,
				.header-fixed > thead > tr,
				.header-fixed > tbody > tr,
				.header-fixed > thead > tr > th,
				.header-fixed > tbody > tr > td {
				    display: block;
				}

				.header-fixed > tbody > tr:after,
				.header-fixed > thead > tr:after {
				    content: ' ';
				    display: block;
				    visibility: hidden;
				    clear: both;
				}

				.header-fixed > tbody {
				    overflow-y: auto;
				    height: 500px;
				}

				.header-fixed > tbody > tr > td,
				.header-fixed > thead > tr > th {
				    width: 16.6%;
				    float: left;
				}


            </style>



    <br>

    <div class="alert alert-info ">
  		You are viewing information relating to sample:  <strong> {{ sample.name }} </strong> 
  		

	</div>

	{% if sample.sample_qc_passed == 'FAIL' %}

	   <div class="alert alert-danger ">
  		 <strong> Danger!  </strong> The sample has failed automatic quality control.


					
	</div>

	{% endif %}


   <ul class="nav nav-tabs">


	    <li class="active"><a data-toggle="tab" href="#variants">Variants</a></li>
	    <li><a data-toggle="tab" href="#filter">Filter</a></li>
	    <li><a data-toggle="tab" href="#QC">QC</a></li>
	   	<li><a data-toggle="tab" href="#coverage">Coverage</a></li>
	    <li><a data-toggle="tab" href="#history">History</a></li>
	    <li><a data-toggle="tab" href="#report">Report</a></li>

  	</ul>

  	<br>

  	<div class="tab-content" >


  	<!--Variants Tab displays filtered variants -->
  	<div id="variants" class="tab-pane active">

  	 <div class ="container-fluid" id='detail'>
		    <ul class="nav nav-pills nav-stacked col-lg-1">
		      <li ><a href="#tab_a" data-toggle="pill">Detail</a></li>
		      <li class="active"><a href="#tab_b" data-toggle="pill">IGV</a></li>
		      <li><a href="#tab_c" data-toggle="pill">Evidence</a></li>
		    </ul>
	    <div class="tab-content col-lg-11">

	    <div class="tab-pane" id="tab_a">
	    <h4>Detail</h4>



		            <table class = 'table table-bordered' >

		            <thead>

		              <tr>
		                <th colspan="2">Summary</th>
		              </tr>

		            </thead>

		            <tbody>

		              <tr>
		                <td> Chromsome </td>
		                <td> {{ variant.chromosome }} </td>

		              </tr>

		              <tr>
		                <td> Position </td>
		                <td> {{ variant.position }} </td>

		              </tr>

		              <tr>
		                <td> Ref </td>
		                <td> {{ variant.ref }} </td>

		              </tr>

		              <tr>
		                <td> Alt </td>
		                <td> {{ variant.alt }} </td>

		              </tr>

		              <tr>
		                <td> ID </td>
		                <td> {{ variant.display_ids|join:", " }} </td>

		              </tr>


		              <tr>
		                <td> Genes </td>
		                <td>

		                {% for gene in variant.get_genes %}


		                  {{ gene }} 


		                {% endfor %}


		                </td>

		              </tr>


		              <tr>
		                <td> Worst Consequence </td>
		                <td> {{ variant.worst_consequence }} </td>

		              </tr>

		              <tr>
		                <td> Rated as Pathogenic </td>
		                <td> {{ variant.rated_as_pathogenic}} </td>

		              </tr>

		              <tr>
		                <td> ClinVar Clin Sig </td>
		                <td> {{ variant.clinical_sig }} </td>

		              </tr>


		              <tr>
		                <td> Picked </td>
		                <td> {{ variant.get_picked_transcript }} </td>

		              </tr>

		            </tbody>
		          </table>



	    </div>

	      <div class="tab-pane active" id="tab_b">

	      <h4>IGV</h4>
	      	





                
            </div>
            <div class="tab-pane" id="tab_c">
                 <h4>Evidence</h4>
                
            </div>
            <div class="tab-pane" id="tab_d">
                 <h4>Comments</h4>
                
            </div>
    </div><!-- tab content -->
    </div><!-- end of container -->


<hr>
        

		<strong> Variants: </strong> {{summary.total}} / {{total_summary.total}} | <strong> Missense:</strong> {{summary.missense_count}} / {{total_summary.missense_count}} | 
	    <strong> Indels </strong> {{summary.indel_count}} / {{total_summary.indel_count}} | <strong> LOF : </strong> {{summary.lof_count}} / {{total_summary.lof_count}} |
	    <strong> Synonymous </strong> {{summary.synonymous}} / {{total_summary.synonymous}}

<hr>



  <table id="variant_table" class = "table table-hover">

          <thead>

        <tr>
          <th style="display:none;"> Variant Hash </th>
          <th> HGVSc </th>
          <th> Variant </th>
          <th> Ref </th>
          <th> Alt </th>
          <th> Worst Conseqeunce </th>
          <th> Genes </th>
          <th> Max AF </th>





        </tr>

      </thead>

      <tbody>


      {% for variant in variants %}

      <tr class='variant'>

        <td style="display:none;"> {{variant.variant_hash}}</td>
        <td> {{variant.get_picked_hgvsc}} </td>

        <td> {{variant.chromosome}} {{variant.position}} </td>
        <td> {{variant.ref}} </td>
        <td> {{variant.alt}} </td>
        <td> {{variant.worst_consequence}} </td>
        <td> {{variant.get_genes|join:" "}} </td>
         <td> {{variant.max_af}} </td>


      </tr>

      {% endfor %}





      </tbody>
      </table>


  	      			
                	

                  	

      </div>

      <!--Filter Tab  -->
	<div id="filter" class="tab-pane fade">

	     <button type="button" class="btn btn-default btn-sm" id='selectall'>
          <span class="glyphicon glyphicon-check"></span> Select All
        </button>

        <hr>



		<div class="alert alert-info">

  	<form class="form-horizontal" method="get" >

            {{ filter_form.as_p }}

        <br>

        <div class="form-actions">

            <button type="submit" class="btn btn-primary" name="filterform">Filter</button>

        </div>
    
        </form>




	</div>



	</div>




	<!--QC Tab with link to QC -->
	<div id="QC" class="tab-pane fade">

	

	<div class="alert alert-info">
  		<strong>QC Status: {{sample.sample_qc_passed}}  </strong> 
	</div>


        <table class = 'table table-bordered' >
	        <thead>
	          <tr>
	            <th colspan="2">Summary</th>

	          </tr>
	        </thead>

			<tbody>

			<tr>

				<td> Raw Total Sequences </td>
				<td> {{sample.raw_total_sequences}} </td>


			</tr>

			<tr>

				<td> Filtered </td>
				<td> {{sample.filtered_sequences}} </td>

			</tr>

			<tr>

				<td> Non-Primary </td>
				<td> {{sample.non_primary_alignments}} </td>

			</tr>

			<tr>

				<td> Duplicated </td>
				<td> {{sample.reads_duplicated}} </td>

			</tr>

			<tr>

				<td> Mapped </td>
				<td> {{sample.reads_mapped}} </td>

			</tr>

			<tr>

				<td> Zero MQ </td>
				<td> {{sample.reads_MQ0}} </td>

			</tr>


			<tr>

				<td> Average Read Length </td>
				<td> {{sample.average_length}} </td>

			</tr>


			<tr>

				<td> Total Bases </td>
				<td> {{sample.total_length}} </td>

			</tr>

			<tr>

				<td> Total Bases Mapped </td>
				<td> {{sample.bases_mapped_cigar}} </td>

			</tr>

			<tr>

				<td> Base Error Rate </td>
				<td> {{sample.get_error_rate|floatformat:"5"}} </td>

			</tr>

			</tbody>


		</table>



		<hr>

		<h3> Sam Stats Figures </h3>

		<hr>
        <table class="imgs">

        <tbody>

			<tr>

			<td>
			{% if sample.acgt_cycles_image %}



			<a class="thumbnail" href={{sample.acgt_cycles_image.url}}><img src={{sample.acgt_cycles_image.url}} width="150px"><span>Per Base Sequence Content<br><img src={{sample.acgt_cycles_image.url}}></span></a>


			{%endif%}

			</td>

			<td>

			{% if sample.coverage_image %}

			<a class="thumbnail" href={{sample.coverage_image.url}}><img src={{sample.coverage_image.url}} width="150px"><span>Coverage<br><img src={{sample.coverage_image.url}}></span></a>

			{%endif%}

			</td>

			<td>

			{% if sample.gc_content_image %}

			<a class="thumbnail" href={{sample.gc_content_image.url}}><img src={{sample.gc_content_image.url}} width="150px"><span>GC Content<br><img src={{sample.gc_content_image.url}}></span></a>

			{%endif%}

			</td>


			</tr>

			<tr>


			<td>

			{% if sample.gc_depth_image %}

			<a class="thumbnail" href={{sample.gc_depth_image.url}}><img src={{sample.gc_depth_image.url}} width="150px"><span>GC Depth<br><img src={{sample.gc_depth_image.url}}></span></a>

			{%endif%}

			 </td>

			<td>

			{% if sample.indel_cycles_image %}

			<a class="thumbnail" href={{sample.indel_cycles_image.url}}><img src={{sample.indel_cycles_image.url}} width="150px"><span>Indel Cycles<br><img src={{sample.indel_cycles_image.url}}></span></a>

			{%endif%}


			</td>

			<td>

			{% if sample.indel_dist_image %}

			<a class="thumbnail" href={{sample.indel_dist_image.url}}><img src={{sample.indel_dist_image.url}} width="150px"><span>Indel Dist<br><img src={{sample.indel_dist_image.url}}></span></a>

			{%endif%}

			</td>

			</tr>

			<tr>

			<td>

			{% if sample.insert_size_image %}

			<a class="thumbnail" href={{sample.insert_size_image.url}}><img src={{sample.insert_size_image.url}} width="150px"><span>Insert Size<br><img src={{sample.insert_size_image.url}}></span></a>

			{%endif%}

			</td>

			<td>

			{% if sample.quality_cycle_image %}

			<a class="thumbnail" href={{sample.quality_cycle_image.url}}><img src={{sample.quality_cycle_image.url}} width="150px"><span>Quality Cycle<br><img src={{sample.quality_cycle_image.url}}></span></a>

			{%endif%}

			</td>

			<td>

			{% if sample.quality_cycle_read_image %}

			<a class="thumbnail" href={{sample.quality_cycle_read_image.url}}><img src={{sample.quality_cycle_read_image.url}} width="150px"><span>Quality Cycle 2<br><img src={{sample.quality_cycle_read_image.url}}></span></a> 

			{%endif%}

			</td>

			</tr>

			<tr>

			<td>

			{% if sample.quality_cycle_read_freq_image %}

			<a class="thumbnail" href={{sample.quality_cycle_read_freq_image.url}}><img src={{sample.quality_cycle_read_freq_image.url}} width="150px"><span>Quality Cycle 2<br><img src={{sample.quality_cycle_read_freq_image.url}}></span></a>

			{%endif%}

			</td>


			<td>

			{% if sample.quality_cycle_read_freq_image %}


			<a class="thumbnail" href={{sample.quality_heatmap_image.url}}><img src={{sample.quality_heatmap_image.url}} width="150px"><span>Quality Cycle 2<br><img src={{sample.quality_heatmap_image.url}}></span></a>

			{%endif%}

			</td>


			</tr>


		

		</tbody>


    </table>


	</div>


	<!--Coverage Tab with Gene/Exon Coverage -->
	<div id="coverage" class="tab-pane fade">

	<h3> Gene Coverage </h3>

	     <button type="button" class="btn btn-default btn-sm" id='togglecovdata'>
          <span class="glyphicon glyphicon-plus"></span> Toggle
        </button>




	<table class = 'table table-hover table-condensed'>



	  	  	<thead>

				<tr>
					<th> Gene </th>
					<th  class ='hidecol' style=display:none> 100x </th>
					<th  class ='hidecol'style=display:none> 200x </th>
					<th  class ='hidecol'style=display:none> 300x </th>
					<th  class ='hidecol'style=display:none> 400x </th>
					<th  class ='hidecol'style=display:none> 500x </th>
					<th  class ='hidecol'style=display:none> 600x </th>
					<th> Min </th>
					<th> Max </th>
					<th> Mean </th>

					<th> pct > 100x </th>
					<th> pct > 200x </th>
					<th> pct > 300x </th>
					<th> pct > 400x </th>
					<th> pct > 500x </th>
					<th> pct > 600x </th>

				</tr>

			</thead>

			<tbody>

			{% for gene in gene_coverage %}

			<tr>

				<td> {{gene.gene}} </td>
				<td  class ='hidecol'style=display:none>  {{gene.x100}} </td>
				<td  class ='hidecol'style=display:none> {{gene.x200}} </td>
				<td  class ='hidecol'style=display:none> {{gene.x300}} </td>
				<td  class ='hidecol'style=display:none> {{gene.x400}} </td>
				<td  class ='hidecol'style=display:none> {{gene.x500}} </td>
				<td  class ='hidecol'style=display:none> {{gene.x600}} </td>
				<td> {{gene.min_coverage}} </td>
				<td> {{gene.max_coverage}} </td>
				<td> {{gene.mean_coverage}} </td>




				{% with gene_percentages=gene.get_percentages %}


				<td> {{gene_percentages.0}} </td>
				<td> {{gene_percentages.1}} </td>
				<td> {{gene_percentages.2}} </td>
				<td> {{gene_percentages.3}} </td>
				<td> {{gene_percentages.4}} </td>
				<td> {{gene_percentages.5}} </td>


				{% endwith %}





			</tr>

			{% endfor %}

			</tbody>

	</table>

	<h3> Exon Coverage </h3>





		<table class = 'table table-hover table-condensed'>

	  	  	<thead>

				<tr>
					<th visibility:hidden> Gene </th>
					<th> Exon </th>
					<th class ='hidecol' style=display:none> 100x </th>
					<th class ='hidecol' style=display:none> 200x </th>
					<th class ='hidecol' style=display:none> 300x </th>
					<th class ='hidecol' style=display:none> 400x </th>
					<th class ='hidecol' style=display:none> 500x </th>
					<th class ='hidecol' style=display:none> 600x </th>
					<th> Min </th>
					<th> Max </th>
					<th> Mean </th>


					<th> pct > 100x </th>
					<th> pct > 200x </th>
					<th> pct > 300x </th>
					<th> pct > 400x </th>
					<th> pct > 500x </th>
					<th> pct > 600x </th>

				</tr>

			</thead>

			<tbody>

			{% for exon in exon_coverage %}

			<tr>

				<td> {{exon.gene}} </td>
				<td> {{exon.exon}} </td>
				<td class ='hidecol' style=display:none> {{exon.x100}} </td>
				<td class ='hidecol' style=display:none> {{exon.x200}} </td>
				<td class ='hidecol' style=display:none> {{exon.x300}} </td>
				<td class ='hidecol' style=display:none> {{exon.x400}} </td>
				<td class ='hidecol' style=display:none> {{exon.x500}} </td>
				<td class ='hidecol' style=display:none> {{exon.x600}} </td>
				<td> {{exon.min_coverage}} </td>
				<td> {{exon.max_coverage}} </td>
				<td> {{exon.mean_coverage}} </td>



				{% with exon_percentages=exon.get_percentages %}

				
				<td> {{exon_percentages.0}} </td>
				<td> {{exon_percentages.1}} </td>
				<td> {{exon_percentages.2}} </td>
				<td> {{exon_percentages.3}} </td>
				<td> {{exon_percentages.4}} </td>
				<td> {{exon_percentages.5}} </td>


				{% endwith %}





			</tr>

			{% endfor %}

			</tbody>

	</table>





	



	</div>














	<!--History Tab with link to History -->
	<div id="history" class="tab-pane fade">

	<p> History </p>



	</div>


	<div id="report" class="tab-pane fade">



			<table class = 'table table-hover'>

	  	  	<thead>

				<tr>
					<th> Report ID </th>
					<th> Creation Date </th>
					<th> Status </th>
					<th> Author </th>
					<th> View </th>



				</tr>

			</thead>

			<tbody>

				{% for report in reports %}

				<tr>

					<td> {{report.pk}} </td>
					<td> {{report.get_creation_date }}</td>
					<td> {{report.get_status }}</td>
					<td> {{ report.get_author}}</td>
					<td> <a href="{% url 'view_sample_report' pk_sample=sample.pk pk_report=report.pk  %}"> View  </a></td>


				</tr>

				{% endfor %}


			</tbody>

		</table>

		<hr>



        <div class="panel panel-success">
        <div class="panel-heading">Reports</div>
        <div class="panel-body"> <form class="form-horizontal" method="post" >{% csrf_token %}

            {{ report_form.as_p }}

        <br>

        <div class="form-actions">

            <button type="submit" class="btn btn-primary" name="reportform">Create New</button>

        </div>


    
        </form>

        </div>
        </div>
        



	</div>
	</div>

</div>


<script>
$(document).ready(function(){
    $(".variant").click(function(){


	//add colour to selected row	
	var variant_hash  = $(this).children().eq(0).text();

	var selected = $(this).hasClass("success");
	$(this).siblings().removeClass("success");


	if(!selected)
	$(this).addClass("success");



      $.ajax({
    url: '/ajax_detail',
    type: 'get', // This is the default though, you don't actually need to always mention it
    data: {"variant_hash" : variant_hash, 'sample_pk': {{sample.pk}}},
    success: function(data) {

        $("#detail").html(data);
    },
    failure: function(data) { 
        alert('Got an error dude');
    }
}); 


    });


    $("#togglecovdata").click(function(){ //toggle coverage columns with only raw data i.e. not percentages

    	$(".hidecol").toggle();

    });

    $("#selectall").click(function(){ //select everything in the filter form


    	$("#id_upstream_gene_variant").prop("checked", true);
		$("#id_transcript_amplification").prop("checked", true);
		$("#id_transcript_ablation").prop("checked", true);
		$("#id_synonymous_variant").prop("checked", true);
		$("#id_stop_retained_variant").prop("checked", true);
		$("#id_stop_lost").prop("checked", true);
		$("#id_stop_gained").prop("checked", true);
		$("#id_start_lost").prop("checked", true);
		$("#id_splice_region_variant").prop("checked", true);
		$("#id_splice_donor_variant").prop("checked", true);
		$("#id_splice_acceptor_variant").prop("checked", true);
		$("#id_regulatory_region_variant").prop("checked", true);
		$("#id_regulatory_region_amplification").prop("checked", true);
		$("#id_regulatory_region_ablation").prop("checked", true);
		$("#id_protein_altering_variant").prop("checked", true);
		$("#id_non_coding_transcript_variant").prop("checked", true);
		$("#id_non_coding_transcript_exon_variant").prop("checked", true);
		$("#id_missense_variant").prop("checked", true);
		$("#id_mature_miRNA_variant").prop("checked", true);
		$("#id_intron_variant").prop("checked", true);
		$("#id_intergenic_variant").prop("checked", true);
		$("#id_inframe_insertion").prop("checked", true);
		$("#id_inframe_deletion").prop("checked", true);
		$("#id_incomplete_terminal_codon_variant").prop("checked", true);
		$("#id_frameshift_variant").prop("checked", true);
		$("#id_feature_truncation").prop("checked", true);
		$("#id_feature_elongation").prop("checked", true);
		$("#id_downstream_gene_variant").prop("checked", true);
		$("#id_coding_sequence_variant").prop("checked", true);
		$("#id_TF_binding_site_variant").prop("checked", true);
		$("#id_TFBS_amplification").prop("checked", true);
		$("#id_TFBS_ablation").prop("checked", true);
		$("#id_NMD_transcript_variant").prop("checked", true);
		$("#id_five_prime_UTR_variant").prop("checked", true);
		$("#id_three_prime_UTR_variant").prop("checked", true);

		$("#id_freq_max_af").val(1);



    });




});
</script>






 
{% endblock %}  
  
  
 

  
  
